05-agent-demo/risingwave-agent.mdx

```diff
@@ -0,0 +1,626 @@
+                usage.input_tokens
+                + getattr(usage, "cache_read_input_tokens", 0)
+                + getattr(usage, "cache_creation_input_tokens", 0)
+            )
+            output_tokens = usage.output_tokens
+
+            current_turn_input = total_input - self.total_tokens
+            self.message_tokens.append((current_turn_input, output_tokens))
+            self.total_tokens += current_turn_input + output_tokens
+
+    def truncate(self) -> None:
+        """Remove oldest messages when context window limit is exceeded."""
+        if self.total_tokens <= self.context_window_tokens:
+            return
+
+        TRUNCATION_NOTICE_TOKENS = 25
+        TRUNCATION_MESSAGE = {
+            "role": "user",
+            "content": [
+                {
+                    "type": "text",
+                    "text": "[Earlier history has been truncated.]",
+                }
+            ],
+        }
+
+        def remove_message_pair():
+            self.messages.pop(0)
+            self.messages.pop(0)
+
+            if self.message_tokens:
+                input_tokens, output_tokens = self.message_tokens.pop(0)
+                self.total_tokens -= input_tokens + output_tokens
+
+        while (
+            self.message_tokens
+            and len(self.messages) >= 2
+            and self.total_tokens > self.context_window_tokens
+        ):
+            remove_message_pair()
+
+            if self.messages and self.message_tokens:
+                original_input_tokens, original_output_tokens = (
+                    self.message_tokens[0]
+                )
+                self.messages[0] = TRUNCATION_MESSAGE
+                self.message_tokens[0] = (
+                    TRUNCATION_NOTICE_TOKENS,
+                    original_output_tokens,
+                )
+                self.total_tokens += (
+                    TRUNCATION_NOTICE_TOKENS - original_input_tokens
+                )
+
+    def format_for_api(self) -> list[dict[str, Any]]:
+        """Format messages for Claude API with optional caching."""
+        result = [
+            {"role": m["role"], "content": m["content"]} for m in self.messages
+        ]
+
+        if self.enable_caching and self.messages:
+            result[-1]["content"] = [
+                {**block, "cache_control": {"type": "ephemeral"}}
+                for block in self.messages[-1]["content"]
+            ]
+        return result
+```
+
+The create_mcp_connection function serves as a factory to create either a stdio or sse MCP connection based on the configuration provided. The setup_mcp_connections async function loops through a list of server configs, creates connections, retrieves each server’s available tools, and dynamically generates async wrapper functions that the agent can later call. It collects both the tool definitions and their corresponding function wrappers into dictionaries for use in the agent loop
+
+```python
+class MCPConnection(ABC):
```

Comment:
I'm not sure if we need this abstraction. Checkout https://gofastmcp.com/clients/client#creating-a-client